qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java

```diff
@@ -87,6 +90,7 @@ public void testSnapshotBasedRecovery() throws Exception {
                     // That is an issue only for the first mixed round.
                     // In that case we exclude the upgraded node from the shard allocation and cancel the shard to force moving
                     // the primary to a node in the old version, this allows adding replicas in the first mixed round.
+                    logger.info("--> Primary node in first mixed round {} / {}", primaryNodeId, primaryNodeVersion);
```

Comment:
I think rebalance can kick in after this check, I'm not sure if there's another way to fix this other than disabling the rebalancing during this test.

Reply:
> It seems there is some leniency in the excluded node setting above though. I think we could strengthen that instead in following ways:

Yes, that seems to be the case, since the first upgraded node is restarted, meaning that if the primary was located in  that node, the replica is promoted. But it seems like sometimes the primary moves back to the upgraded node. I've pushed the changes that you suggested. Let's see if that helps.